erpnext/regional/doctype/import_supplier_invoice/import_supplier_invoice.py

```diff
@@ -0,0 +1,386 @@
+from bs4 import BeautifulSoup as bs
+from frappe.utils import cint, flt, today, nowdate, add_days, get_files_path
+import dateutil
+from frappe.utils.file_manager import save_file
+
+class ImportSupplierInvoice(Document):
+
+	def autoname(self):
+		if not self.name:
+			self.name = "Import Invoice on " + format_datetime(self.creation)
+
+	def import_xml_data(self):
+		import_file = frappe.get_doc("File", {"file_url": self.zip_file})
+		self.publish("File Import", _("Processing XML Files"), 1, 3)
+
+		pi_count = 0
+		mop_options = frappe.get_meta('Mode of Payment').fields[4].options
+		mop_str = re.sub('\n', ',', mop_options)
+		mop_dict = dict(item.split("-") for item in mop_str.split(","))
+
+		with zipfile.ZipFile(get_full_path(self.zip_file)) as zf:
+			file_count = 0
+			for file_name in zf.namelist():
+				items = []
+				taxes = []
+				terms = []
+				encoded_content = zf.read(file_name)
+
+				try:
+					content = encoded_content.decode("utf-8-sig")
+				except UnicodeDecodeError:
+					try:
+						content = encoded_content.decode("utf-16")
+					except UnicodeDecodeError as e:
+						frappe.log_error(message=e, title="UTF-16 encoding error for File Name: " + file_name)
+
+				file_content = bs(content, "xml")
+
+				for line in file_content.find_all("DatiTrasmissione"):
+					destination_code = line.CodiceDestinatario.text
+
+				for line in file_content.find_all("DatiGeneraliDocumento"):
+					document_type = line.TipoDocumento.text
+					bill_date = dateutil.parser.parse(line.Data.text).strftime("%Y-%m-%d")
+					invoice_no = line.Numero.text
+					if len(invoice_no) != 0:
+						for line in file_content.find_all("CedentePrestatore"):
+							tax_id = line.DatiAnagrafici.IdPaese.text + line.DatiAnagrafici.IdCodice.text
+							if line.find("CodiceFiscale"):
+								fiscal_code = line.DatiAnagrafici.CodiceFiscale.text
+							else:
+								fiscal_code = ""
+							if line.find("RegimeFiscale"):
+								fiscal_regime = line.DatiAnagrafici.RegimeFiscale.text
+							else:
+								fiscal_regime = ""
+							if line.find("Denominazione"):
+								supplier = line.DatiAnagrafici.Anagrafica.Denominazione.text
+							if line.find("Nome"):
+								supplier = line.DatiAnagrafici.Anagrafica.Nome.text + " " + line.DatiAnagrafici.Anagrafica.Cognome.text
+							address_line1 = line.Sede.Indirizzo.text
+							city = line.Sede.Comune.text
+							if line.find("Provincia"):
+								province = line.Sede.Provincia.text
+							else:
+								province = ""
+							pin_code = line.Sede.CAP.text
+							country = get_country(line.Sede.Nazione.text)
+
+						total_discount = 0
+
+						for line in file_content.find_all("DettaglioLinee"):
+							if line.find("PrezzoUnitario") and line.find("PrezzoTotale"):
+								unit_rate = float(line.PrezzoUnitario.text) or float(0)
+								line_total = float(line.PrezzoTotale.text) or float(0)
+
+								if (unit_rate == 0.0):
+									qty = float(1)
+									uom = "nos"
```

Comment:
Why hardcode? maybe we can use "Default Stock UOM" set in the Stock Settings. If Default Stock UOM is blank then throw error

Reply:
Ok